lib/kube/proxy/forwarder.go

```diff
@@ -2127,32 +2196,42 @@ type clusterSession struct {
 	// rbacSupportedResources is the list of resources that support RBAC for the
 	// current cluster.
 	rbacSupportedResources rbacSupportedResources
-	// connCtx is the context used to monitor the connection.
-	connCtx context.Context
-	// connMonitorCancel is the conn monitor connMonitorCancel function.
-	connMonitorCancel context.CancelCauseFunc
+
+	// mu protects access to the monitorCancel
+	mu sync.Mutex
+	// monitorCancel is the conn monitor monitorCancel function.
+	monitorCancel context.CancelCauseFunc
 }
 
 // close cancels the connection monitor context if available.
 func (s *clusterSession) close() {
-	if s.connMonitorCancel != nil {
-		s.connMonitorCancel(io.EOF)
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.monitorCancel != nil {
+		s.monitorCancel(io.EOF)
+		s.monitorCancel = nil
 	}
 }
 
 func (s *clusterSession) monitorConn(conn net.Conn, err error) (net.Conn, error) {
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-
+	s.mu.Lock()
```

Comment:
why do we need to cancel the ctx and redo it?

Reply:
I've rolled back this change, since now we don't have fallback executor with 2 possible connection attempts.